### PR TITLE
firefox: fix cross compilation

### PR DESCRIPTION
**Pull request by NickCao**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Mindavi** [comment]
Please mention what you tested here.

**Comment by NickCao** [comment]
before:
```
nix-repl> pkgsCross.riscv64.rustc.llvmPackages.stdenv.buildPlatform.config
"x86_64-unknown-linux-gnu"

nix-repl> pkgsCross.riscv64.rustc.llvmPackages.stdenv.hostPlatform.config
"x86_64-unknown-linux-gnu"

nix-repl> pkgsCross.riscv64.rustc.llvmPackages.stdenv.targetPlatform.config
"riscv64-unknown-linux-gnu"
```

after:
```
nix-repl> pkgsCross.riscv64.rustc.llvmPackages.stdenv.buildPlatform.config
"x86_64-unknown-linux-gnu"

nix-repl> pkgsCross.riscv64.rustc.llvmPackages.stdenv.hostPlatform.config
"riscv64-unknown-linux-gnu"

nix-repl> pkgsCross.riscv64.rustc.llvmPackages.stdenv.targetPlatform.config
"riscv64-unknown-linux-gnu"
```

**Comment by NickCao** [comment]
It should have no effect on native builds, just testing cross builds.

**Comment by Mindavi** [comment]
Btw, this PR added it: https://github.com/NixOS/nixpkgs/pull/144219

**Comment by NickCao** [comment]
`rebuild: 0` is a nice indication, just found that `lld` fails to cross, fixing on my side.

**Comment by NickCao** [comment]
builds are all green: https://hydra.nichi.co/eval/39, lto on riscv is expected to fail as it is not implemented by lld.

File: pkgs/applications/networking/browsers/firefox/common.nix [hunk 1]
```diff
@@ -196,7 +214,9 @@ buildStdenv.mkDerivation ({
       rustc
       which
       unzip
+      zip
       wrapGAppsHook
+      gnum4
```
Comment (Mindavi) [inline pkgs/applications/networking/browsers/firefox/common.nix:219]: Nit: alphabetic order?

File: pkgs/applications/networking/browsers/firefox/common.nix [hunk 2]
```diff
@@ -130,7 +140,10 @@ buildStdenv.mkDerivation ({
   inherit src unpackPhase meta;
 
   patches = [
-  ] ++
+  ] ++ lib.optional ((lib.versionAtLeast version "94") && buildStdenv.hostPlatform.isRiscV) (fetchpatch {
```
Comment (Mindavi) [inline pkgs/applications/networking/browsers/firefox/common.nix:143]: This patch is rather big, but I guess that's ok. Could it be applied unconditionally? Is it (being) upstreamed and do you have a link?
(Just some questions, but always applying is generally preferred if possible.)

File: pkgs/applications/networking/browsers/firefox/common.nix [hunk 3]
```diff
@@ -116,8 +126,8 @@ let
   # doesn't work on it. For now it is fine since ltoSupport implies no Darwin.
   buildStdenv = if ltoSupport
                 # LTO requires LLVM bintools including ld.lld and llvm-ar.
-                then overrideCC llvmPackages.stdenv (llvmPackages.stdenv.cc.override {
-                  inherit (llvmPackages) bintools;
+                then overrideCC llvmPackagesStdenv.stdenv (llvmPackagesStdenv.stdenv.cc.override {
```
Comment (Mindavi) [inline pkgs/applications/networking/browsers/firefox/common.nix:129]: Is stdenv.stdenv needed?

File: pkgs/applications/networking/browsers/firefox/common.nix (hunk 2 above)
Comment (NickCao) [reply pkgs/applications/networking/browsers/firefox/common.nix:143]: It could be applied unconditionally, but I think it's not worth the trouble for dealing with possible regressions. The upstream issue is here: https://bugzilla.mozilla.org/show_bug.cgi?id=1717203

File: pkgs/applications/networking/browsers/firefox/common.nix (hunk 3 above)
Comment (NickCao) [reply pkgs/applications/networking/browsers/firefox/common.nix:129]: llvmPackagesStdenv is actually a name for llvmPackages *for* Stdenv, not a stdenv

**Comment by NickCao** [comment]
builds fine natively on my local machine, we should be good to go.

**Review by Mindavi — APPROVED** [review]
Diff LGTM now

File: pkgs/applications/networking/browsers/firefox/common.nix (hunk 2 above)
Comment (symphorien) [reply pkgs/applications/networking/browsers/firefox/common.nix:143]: could you put the link in the nix file?

**Comment by NickCao** [comment]
Dropped the riscv related patches as they no longer apply as of firefox 95, but build for other platforms should work.

**Comment by Mic92** [comment]
What cross compilation target did you test? aarch64 does not seem to cross compile for me.

**Comment by NickCao** [comment]
Also aarch64, just ran a build, and it was green: https://hydra.nichi.co/build/82010.

**Comment by Mindavi** [comment]
It might be good to list which PRs (if they're already open) are blocking this to work then. It's not _that_ useful to merge this if more work is needed, but I guess it's debatable. At least it's hard to test if it doesn't work ootb.

**Comment by NickCao** [comment]
We now depends on https://github.com/NixOS/nixpkgs/pull/148647 (staging), https://github.com/NixOS/nixpkgs/pull/148618 (WIP), and https://github.com/NickCao/nixpkgs/commit/851d3bbb1e7dbffe094c6b758a5447045146daa7

**Comment by lovesegfault** [comment]
Switch to `libvpx` and test, I suspect it will be fine but it's hard to know. Looking at what other distros (e.g. Gentoo) do is also valuable.

**Comment by NickCao** [comment]
> Switch to `libvpx` and test, I suspect it will be fine but it's hard to know. Looking at what other distros (e.g. Gentoo) do is also valuable.

I've check that both Gentoo and Debian are using the latest libvpx for firefox. Do require additional testing on our side.

**Comment by Mic92** [comment]
> > Switch to `libvpx` and test, I suspect it will be fine but it's hard to know. Looking at what other distros (e.g. Gentoo) do is also valuable.
> 
> I've check that both Gentoo and Debian are using the latest libvpx for firefox. Do require additional testing on our side.

Maybe just open a PR for this and I can build all firefox at some point.

File: pkgs/applications/networking/browsers/firefox/common.nix [hunk 4]
```diff
@@ -245,6 +262,8 @@ buildStdenv.mkDerivation ({
     export BUILD_OFFICIAL=1
   '');
 
+  configurePlatforms = [ ];
```
Comment (lovesegfault) [inline pkgs/applications/networking/browsers/firefox/common.nix:265]: Why is this needed?

Comment (NickCao) [reply pkgs/applications/networking/browsers/firefox/common.nix:265]: https://github.com/NixOS/nixpkgs/blob/640239d7d6414d98c0ab286d76d6bd8962df938b/pkgs/applications/networking/browsers/firefox/common.nix#L285

Comment (ius) [reply pkgs/applications/networking/browsers/firefox/common.nix:265]: Should add a comment to explain, otherwise it'll end up being removed by accident, see https://github.com/NixOS/nixpkgs/pull/152715

**Comment by Mindavi** [comment]
The gtk3 patchset is merged, hopefully that'll be enough to get this working. Could you rebase this? Or do you want to wait until the gtk3 change hits master before investing more time in this (the build will be very long for sure for firefox, I'm guessing).

**Comment by NickCao** [comment]
The spring cleanup changed quite a lot of things, it will take me some time to pick this up.

**Comment by NickCao** [comment]
Cross builds fine with LTO for aarch64-linux.

File: pkgs/top-level/all-packages.nix [hunk 5]
```diff
@@ -12690,7 +12690,7 @@ with pkgs;
   gcc_latest = gcc11;
 
   # Use the same GCC version as the one from stdenv by default
-  gfortran = wrapCC (gccStdenv.cc.cc.override {
+  gfortran = wrapCC (gcc11.cc.override {
```
Comment (NickCao) [inline pkgs/top-level/all-packages.nix:12693]: Some context in https://github.com/NixOS/nixpkgs/pull/148539, need to find a much more proper fix.

Comment (r-burns) [reply pkgs/top-level/all-packages.nix:12693]: Instead of hardcoding `gcc11`, can we use `gcc` here? My intention was to make the unversioned gfortran attr resolve to the same version as the unversioned gcc attr.

Comment (NickCao) [reply pkgs/top-level/all-packages.nix:12693]: Let me do a rebuild to figure it out.

**Comment by Mindavi** [comment]
If we want to get this in as it is, this needs to go to staging. Since the gtk3 change is only in staging anyway, maybe it makes most sense to just send this all to staging for now.

It's also possible to factor out the gfortran change and send the rest to master, but it'll only work after staging goes to master anyway.

**Comment by NickCao** [comment]
I've changed the base branch to staging, and the gfortran commit can be dropped once https://github.com/NixOS/nixpkgs/pull/171466 is merged.

**Comment by Mindavi** [comment]
Ah, I know why it suddenly did that (show thousands of rebuilds), because gcc11 is a difference from how it was before for some platforms. Now master should be fine again :). Sorry for the churn.

**Comment by NickCao** [comment]
> Ah, I know why it suddenly did that (show thousands of rebuilds), because gcc11 is a difference from how it was before for some platforms. Now master should be fine again :). Sorry for the churn.

Without the gtk fix this would just be useless, going through staging is fine

**Comment by Mindavi** [comment]
Of course, forgot about that. Yes, that's fine.

File: pkgs/applications/networking/browsers/firefox/common.nix [hunk 6]
```diff
@@ -141,16 +142,21 @@ let
...
   };
+  llvmPackagesBuildBuild = llvmPackagesBuildBuild0.override {
+    bootBintoolsNoLibc = null;
+    bootBintools = null;
+  };
 
   # LTO requires LLVM bintools including ld.lld and llvm-ar.
   buildStdenv = overrideCC llvmPackages.stdenv (llvmPackages.stdenv.cc.override {
-    inherit (llvmPackages) bintools;
+    bintools = if ltoSupport then buildPackages.rustc.llvmPackages.bintools else stdenv.cc.bintools;
```
Comment (lovesegfault) [inline pkgs/applications/networking/browsers/firefox/common.nix:159]: Why isn't it always `llvmPackagesBuildBuild.bintools`?

Comment (NickCao) [reply pkgs/applications/networking/browsers/firefox/common.nix:159]: llvm bintools are having trouble creating binaries for riscv, namely linker relaxation is not implemented, let alone LTO. This change ensures llvm bintools are only used when required as in LTO, to workaround such troubles.

**Comment by Madouura** [comment]
I'm not really sure why I was requested to review as I haven't had anything to do with either of these packages.
That said, I'm still a while away from being able to effectively test firefox on riscv64, since right now a basic successful ``nixos-rebuild`` is what I'm aiming for.
I'll see if I can properly test this (and I guess X, some window manager, and networkmanager) once that's complete. Current hardware is pretty slow on riscv, unfortunately.

**Comment by NickCao** [comment]
@Madouura simply running the resulting binary under qemu-user is enough, also I tried watching video with firefox on a hifive unmatched a while ago. it was playing fine despite a little laggy (or a lot, in today's standard). Anyway riscv support is not part of this pr as it requires additional patching.

**Comment by Madouura** [comment]
>Anyway riscv support is not part of this pr as it requires additional patching.

I'm doubly confused then haha.
Anyway, I'll go ahead and do just that then.

**Comment by NickCao** [comment]
> > Anyway riscv support is not part of this pr as it requires additional patching.
> 
> I'm doubly confused then haha. Anyway, I'll go ahead and do just that then.

You can simply grab the patch from our Arch Linux friends: https://github.com/felixonmars/archriscv-packages/tree/master/firefox, disable LTO and get a working firefox out of the box.

**Comment by Mindavi** [comment]
Good to go?

**Comment by Madouura** [comment]
Ah, it's sndio. Makes sense.
The change there is fine.
Gonna make sure x86_64 works and then do the same for aarch64 and approve if nothing is amiss.

**Comment by Madouura** [comment]
x86_64 builds fine.
aarch64-multiplatform fails.
[log.txt](https://github.com/NixOS/nixpkgs/files/8684877/log.txt)

**Comment by NickCao** [comment]
> 

The wrapper won't cross build due to issues with perl, try just firefox-unwrapped.

**Review by Madouura — APPROVED** [review]
LGTM
